### DISCUSSION TITLE
Running pyp-train after classification

### DISCUSSION THREAD

**Discussion by JuhaHuiskonen** [nextpyp/.github#55] [Q&A] [answered]
Hello, could you please help me to figure out how to do the following:  I have picked particles using `pyp-train`.  I have now refined the particles and will classify them next.  To improve the training, I would like to use the particles from selected classes in the training set.  How do I give these positions and run `pyp-train` again?  Do I need to make a copy of the entire pre-processing block?

**Comment by yezhou867** [comment, marked as answer]
Hi, right now nextPYP can not do that way completely from the web interface,  it will involve manipulating some internal files.  We could add some strategies in future updates. 
To make it work now, I will assume you can access the project folders. 
After class selection using the particle filter job, go to the job path and check `csp/ `folder, the `*_boxes3d.txt` should already be updated by changing the last column with `Yes/No` which corresponds to the particle selection as `Yes`. If you count all the lines with `Yes` from all the `boxes3d.txt `files, it should match the particle number given by the filter job log. 
Now you can pick one or two `boxes3d.txt` files and extract the `X, Y, Z` columns from the `Yes `lines to new box files. Be aware that the boxes3d.txt coordinates are in raw image sizes, need to bin them manually according to the binning of the tomograms. Those new box files can be used as input for a new preprocessing job with importing coordinates.  
Once the new preprocessing job finishes, go to the job and view the imported positions, if they look as good as expected,  copy the particle list as a new name which you can use for the new training job.  We never test this way thoroughly, will make sure everything works fine for the new versions.  
<img width="598" alt="Screenshot 2024-02-03 at 7 48 01 PM" src="https://github.com/nextpyp/.github/assets/139471026/8f1d7f1a-b3ba-44f5-94c1-f90f17cce71c">

**Reply by yezhou867** [reply]
Forgot to mention, the `import particles coordinates` parameter needs `.spk` files which are IMOD `.mod` files by converting the box txt using the `point2model` command, make sure the coordinates are in x-y-z order when checking with` imodinfo -a` command.  Swap the order using `imodtrans -Y -T` if necessary.

**Comment by JuhaHuiskonen** [comment]
If I understand correctly, this will require running a copy of the Pre-processing job, including tilt series alignment, CTF etc?  Would it make sense to separate particle picking and extraction to its own block? This way one could test different picking strategies without the need of going back to the Pre-processing job. I am still trying to understand the logic of nextPYP so maybe this doesn't make sense. I suppose I am used to more directional workflows (possibly with several branches, but never going back to the previous block/job to make changes there to do something new - hope this makes sense?).

**Reply by abartesaghi** [reply]
You don't need to create a copy of the pre-processing block in order to use a new set of particles. When you pick particles _manually_ you first need to create a list, so those are saved automatically. If your particles were calculated by nextPYP and you want to keep them, you will need to `Copy` the resulting list of particles and give it a name so you can `Load` them later.

**Reply by JuhaHuiskonen** [reply]
I have figured out how to create lists of particles now.  If I used nextPYP picking, I will make a copy of Auto picked particles and give it a descriptive name.  If I want to pick, let's say ribosomes and segments of microtubules (from lamella) to separate sets, can I then run different Particle refinement blocks in parallel?  How do I define different sets of input particles in Particle refinement?  The Particle refinement block has an option only for Filter tilt series but not Particle list.

**Reply by yezhou867** [reply]
Running separate sets of particles needs 2 branches of preprocessing block and doing them as independent procedures.

**Reply by abartesaghi** [reply]
If you want to avoid re-running the pre-processing block, you can `Import Data` -> `Tomography (from PYP)` and select the path to the existing pre-processing block in the `Path to existing CLI project` field. This will create a new pre-processing block with all the metadata inherited from the existing block. You will still need to `Run` the new block to populate the database, but it won't actually redo any of the calculations. Whether you do it this way or by simply Copying the pre-processing block, note that your Particle Lists will _not_ be transferred over to the new block. However, if particles were  picked by nextPYP in the original pre-processing block, then these _will_ be transferred over.

**Reply by JuhaHuiskonen** [reply]
I created a new project and imported the pre-processing block. I `Run` this block.  Now I have picked manually particles and saved them in a new list.  However, when I want to train a particle picking model `pyp-train`, there's no option (drop down menu on top of the window) to import this set of particles. Suggestions?

**Reply by abartesaghi** [reply]
Indeed, this is a bug, thanks for noticing it! We will fix this in the next patch release.

**Reply by abartesaghi** [reply]
This issue is addressed in [v0.6.3](https://nextpyp.app/files/pyp/0.6.3/docs/index.html).
